cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -188,6 +201,81 @@ public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile
+   * we deserialize the request body into parameter type and pass it when invoking the plugin method and also
+   * optionally pass {@link EndpointPluginContext} to the method if its declared in the plugin method.
+   * @param pluginInstantiator to instantiate plugin instances.
+   * @param pluginClass call having the plugin method to invoke
+   * @param methodParameter parameter body
+   * @param endpointPluginContext endpoint plugin context that can optionally be passed to the method
+   * @throws IOException if there was an exception opening the jar file
+   * @throws InvalidArtifactException if the artifact is invalid
+   * @throws NotFoundException Not Found exception thrown from plugin method invocation or if no matching plugin found.
+   * @throws InvocationTargetException exception thrown from plugin method invocation.
+   */
+  public String invokePluginEndpoint(PluginInstantiator pluginInstantiator, Id.Artifact artifact,
+                                     PluginClass pluginClass, String methodName, String methodParameter,
+                                     DefaultEndpointPluginContext endpointPluginContext)
+    throws IOException, BadRequestException, NotFoundException, ClassNotFoundException, InvocationTargetException {
+
+    // See if there are export packages. Plugins should be in those packages
+
+    ClassLoader pluginClassLoader = pluginInstantiator.getArtifactClassLoader(artifact.toArtifactId());
+
+    Class pluginClassLoaded = pluginClassLoader.loadClass(pluginClass.getClassName());
+
+    Method[] methods = pluginClassLoaded.getDeclaredMethods();
+    for (Method method : methods) {
+      Path pathAnnotation = method.getAnnotation(Path.class);
+      // method should have path annotation else continue
+      if (pathAnnotation == null || !pathAnnotation.value().equals(methodName)) {
+        continue;
+      }
+
+      // make configField name as null, as we might want to call these methods with incomplete config.
+      PluginClass pluginClassWithoutConfig = new PluginClass(pluginClass.getType(), pluginClass.getName(),
+                                                             pluginClass.getDescription(), pluginClass.getClassName(),
+                                                             null, pluginClass.getProperties(),
+                                                             pluginClass.getEndpoints());
+      co.cask.cdap.api.plugin.Plugin plugin =
+        new co.cask.cdap.api.plugin.Plugin(artifact.toArtifactId(), pluginClassWithoutConfig,
+                                           PluginProperties.builder().build());
+      try {
+        Object result;
+        if (method.getParameterTypes().length == 2 &&
+          method.getParameterTypes()[1].equals(EndpointPluginContext.class)) {
+          result = method.invoke(pluginInstantiator.newInstance(plugin),
+                                 GSON.fromJson(methodParameter, method.getParameterTypes()[0]),
```

Comment:
Instead of doing GSON here (and for the result as well), do it in the http handler layer. Serializable is part of the HTTP contract, not the method invocation contract.